lib/logthrdestdrv.c

```diff
@@ -187,74 +270,27 @@ _perform_inserts(LogThreadedDestDriver *self)
 
   while (G_LIKELY(!self->under_termination) &&
          !self->suspended &&
-         (msg = log_queue_pop_head(self->queue, &path_options)) != NULL)
+         (msg = log_queue_pop_head(self->worker.queue, &path_options)) != NULL)
     {
       msg_set_context(msg);
       log_msg_refcache_start_consumer(msg, &path_options);
 
+      self->batch_size++;
       ScratchBuffersMarker mark;
       scratch_buffers_mark(&mark);
       result = self->worker.insert(self, msg);
       scratch_buffers_reclaim_marked(mark);
 
-      switch (result)
-        {
-        case WORKER_INSERT_RESULT_DROP:
-          msg_error("Message dropped while sending message to destination",
-                    evt_tag_str("driver", self->super.super.id));
-
-          _drop_message(self, msg);
-          _disconnect_and_suspend(self);
-          break;
-
-        case WORKER_INSERT_RESULT_ERROR:
-          self->retries.counter++;
-
-          if (self->retries.counter >= self->retries.max)
-            {
-              if (self->messages.retry_over)
-                self->messages.retry_over(self, msg);
-
-              msg_error("Multiple failures while sending message to destination, message dropped",
-                        evt_tag_str("driver", self->super.super.id),
-                        evt_tag_int("number_of_retries", self->retries.max));
-
-              _drop_message(self, msg);
-            }
-          else
-            {
-              _rewind_message(self, msg);
-              _disconnect_and_suspend(self);
-            }
-          break;
-
-        case WORKER_INSERT_RESULT_NOT_CONNECTED:
-          _rewind_message(self, msg);
-          _disconnect_and_suspend(self);
-          break;
-
-        case WORKER_INSERT_RESULT_REWIND:
-          _rewind_message(self, msg);
-          break;
-
-        case WORKER_INSERT_RESULT_SUCCESS:
-          stats_counter_inc(self->written_messages);
-          _accept_message(self, msg);
-          break;
-
-        default:
-          break;
-        }
+      _process_result(self, result);
 
+      log_msg_unref(msg);
       msg_set_context(NULL);
       log_msg_refcache_stop();
     }
-  if (!self->suspended)
+  if (!self->suspended && self->batch_size > 0 && self->worker.flush)
```

Comment:
Until the send in language bindings return with worker_insert_result_t, we cannot add batch_size > 0, as there is no way to do batching after the change, and we break the only way for batching prior the patchset: send returns true, then on-empty is called unconditionally.

Reply:
good catch. I would remove the batch_size condition from here, that does not affect my flush logic that much, it just has to honour the batch_size properly (e.g. do nothing if batch_size is zero). It is not a high price to pay for compatibility.

Once all bindings are migrated, we can put this condition back here.